clang/lib/Basic/Diagnostic.cpp

```diff
@@ -477,6 +486,109 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
       setSeverity(Diag, Map, Loc);
 }
 
+namespace {
+class WarningsSpecialCaseList : public llvm::SpecialCaseList {
+public:
+  static std::unique_ptr<WarningsSpecialCaseList>
+  create(const llvm::MemoryBuffer &MB, std::string &Err) {
+    auto SCL = std::make_unique<WarningsSpecialCaseList>();
+    if (!SCL->createInternal(&MB, Err))
+      return nullptr;
+    return SCL;
+  }
+
+  // Section names refer to diagnostic groups, which cover multiple individual
+  // diagnostics. Expand diagnostic groups here to individual diagnostics.
+  // A diagnostic can have multiple diagnostic groups associated with it, we let
+  // the last section take precedence in such cases.
+  void processSections(DiagnosticsEngine &Diags) {
+    // Drop the default section introduced by special case list, we only support
+    // exact diagnostic group names.
+    Sections.erase("*");
+    // Make sure we iterate sections by their line numbers.
+    std::vector<std::pair<unsigned, const llvm::StringMapEntry<Section> *>>
+        LineAndSectionEntry;
+    LineAndSectionEntry.reserve(Sections.size());
+    for (const auto &Entry : Sections) {
+      LineAndSectionEntry.emplace_back(
+          Entry.second.SectionMatcher->Globs.at(Entry.first()).second, &Entry);
+    }
+    llvm::sort(LineAndSectionEntry);
+    static constexpr auto kFlavor = clang::diag::Flavor::WarningOrError;
+    for (const auto &[_, SectionEntry] : LineAndSectionEntry) {
+      SmallVector<diag::kind, 256> GroupDiags;
+      llvm::StringRef DiagGroup = SectionEntry->getKey();
+      if (Diags.getDiagnosticIDs()->getDiagnosticsInGroup(kFlavor, DiagGroup,
+                                                          GroupDiags)) {
+        StringRef Suggestion =
+            DiagnosticIDs::getNearestOption(kFlavor, DiagGroup);
+        Diags.Report(diag::warn_unknown_diag_option)
+            << static_cast<unsigned>(kFlavor) << DiagGroup
+            << !Suggestion.empty() << Suggestion;
+        continue;
+      }
+      for (diag::kind D : GroupDiags)
+        DiagToSection[D] = &SectionEntry->getValue();
+    }
+  }
+
+  bool isDiagSuppressed(diag::kind D, llvm::StringRef FilePath) const {
+    auto Section = DiagToSection.find(D);
```

Comment:
This is actually an iterator to a pair where the second value is a pointer to Section, which, AFAIU, is declared in http://google3/third_party/llvm/llvm-project/clang/include/clang/Basic/Attr.td;l=2983;rcl=686934080.
If we could move some of the autos in this function to be named types, it might be easier to follow.

Reply:
hmm not sure how you ended up with a reference to that. but no, this is the [Section](https://github.com/llvm/llvm-project/blob/4b44639a4320f980b3c9fa3b96e911e0741f179c/llvm/include/llvm/Support/SpecialCaseList.h#L132) in question.

underlying type is a `DenseMapIterator<diag::kind, llvm::SpecialCaseList::Section*>` modulo some const qualifiers here and there, probably. i haven't really seen any code that actually spells out iterator types, and style guide is also encouraging use of auto in such cases, https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable. so i'd rather keep it as-is

Reply:
Thanks!
I see the benefit of using auto, but in this method it's really hard to follow the types and they don't seem "obvious from the context".
Consider introducing more variables with proper names (like one for Section->second or SrcEntries->second) which would not be auto.
Also, perhaps don't use auto for SrcEntries?

Anyways, "Section" could be rename to something that represents the fact this points to (kind, section).